src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/AzureCredentialsProvider.java

```diff
@@ -0,0 +1,109 @@
+import com.cloudbees.plugins.credentials.CredentialsStore;
+import com.cloudbees.plugins.credentials.common.IdCredentials;
+import com.google.common.base.Suppliers;
+import com.microsoft.azure.PagedList;
+import com.microsoft.azure.keyvault.KeyVaultClient;
+import com.microsoft.azure.keyvault.authentication.KeyVaultCredentials;
+import com.microsoft.azure.keyvault.models.CertificateBundle;
+import com.microsoft.azure.keyvault.models.CertificateItem;
+import com.microsoft.azure.keyvault.models.SecretBundle;
+import com.microsoft.azure.keyvault.models.SecretItem;
+import com.microsoft.jenkins.keyvault.SecretCertificateCredentials;
+import com.microsoft.jenkins.keyvault.SecretStringCredentials;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import hudson.Extension;
+import hudson.model.ItemGroup;
+import hudson.model.ModelObject;
+import hudson.security.ACL;
+import hudson.util.Secret;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.GlobalConfiguration;
+import jenkins.model.Jenkins;
+import org.acegisecurity.Authentication;
+
+
+@Extension
+public class AzureCredentialsProvider extends CredentialsProvider {
+    private static final Logger LOG = Logger.getLogger(AzureCredentialsProvider.class.getName());
+
+    private final AzureCredentialsStore store = new AzureCredentialsStore(this);
+
+    private final Supplier<Collection<IdCredentials>> credentialsSupplier =
+            memoizeWithExpiration(AzureCredentialsProvider::fetchCredentials, Duration.ofMinutes(5));
+
+    private static <T> Supplier<T> memoizeWithExpiration(Supplier<T> base, Duration duration) {
+        return Suppliers.memoizeWithExpiration(base::get, duration.toMillis(), TimeUnit.MILLISECONDS)::get;
+    }
+
+    @NonNull
+    @Override
+    public <C extends Credentials> List<C> getCredentials(@NonNull Class<C> aClass, @Nullable ItemGroup itemGroup,
+                                                          @Nullable Authentication authentication) {
+        if (ACL.SYSTEM.equals(authentication)) {
+            final ArrayList<C> list = new ArrayList<>();
+            for (IdCredentials credential : credentialsSupplier.get()) {
+                if (aClass.isAssignableFrom(credential.getClass())) {
+                    // cast to keep generics happy even though we are assignable..
+                    list.add(aClass.cast(credential));
+                }
+                LOG.log(Level.FINEST, "getCredentials {0} does not match", credential.getId());
+            }
+            return list;
+        }
+
+        return Collections.emptyList();
+    }
+
+    private static Collection<IdCredentials> fetchCredentials() {
```

Comment:
This seems very slow on large vaults is it worth fetching async ?

Reply:
To give you an idea of how bad it is for the vault I'm pointing at:
![image](https://user-images.githubusercontent.com/21194782/72843265-16955a80-3c92-11ea-94ae-1295e67e83b9.png)

Reply:
or is there a less expensive API call? this seems to be retrieving all the secrets, can't we get everything we need from just a list call?